_posts/2016-11-14-JS-Glossary.md

```diff
@@ -0,0 +1,554 @@
+
+### ES6
+
+JavaScript is run by a committee. Around the time that people were starting to talk about HTML5 and CSS3, work was started on a new specification for JavaScript called ECMAScript 6. [What IS ECMA?]
+
+ES6 represents the first point at which JavaScript really started to take a lot of the best features from transpile to JavaScript languages like CoffeeScript. Making it feasible for larger systems programming to be possible in vanilla JavaScript.  
+
+### ES2016
+
+It took forever for [ES6](#es6) to come out, and every time they created / amended a specification there were multiple implementations of the specification available for transpiling via [babel](#babel). This I can imagine was frustrating for developers wanting to use new features, and specification authors trying to put out documentation for discussion as a work in progress. This happened a lot [with the Promises](#promises) API.
+
+To fix this they opted to discuss specification features on a year basis. So that specifications could be smaller and more focused, instead of major multi-year projects. Quite a SemVer jump from 6 to 2016.
+
+### Stages
+
+Turns out that didn't work out too well, so the terminology changed again. The change is mainly to set expectations between the Specification authors and developers transpiling those specifications into their apps.
+
+Now an ECMAScript language improvement specification moves through a series of stages, depending on their maturity. I believe starting at 4, and working down to 1. So a ECMAScript Stage 4 feature is going to be really new, if you're using it via a transpiler then you should expect a lot of potential API changes and code churn. The lower the number, the longer the spec has been discussed, and the more likely for the code you're transpiling to be the vanilla JavaScript code in time.
+
+
+### Modules / Imports
+
+A modules is the terminology for a group of JavaScript code. Terminology can get confusing, as the import structure for a library is very similar to importing a local file.
+
+You can import a module using syntax like `import { thin, other } from "thingy"`. Here's some examples [from our project][imports]:
+
+```js
+// Import modules
+import Relay from 'react-relay'
+import React from 'react'
+// Import two items from the react-native module 
+import { View, TouchableWithoutFeedback } from 'react-native'
+
+// Import the default class from a local file
+import ImageView from '../../opaque_image_view'
+import SwitchBoard from '../../../native_modules/switch_board'
+```
+
+An import can either have [a default export][default-export], or a set of exportable function/objects.
+
+You might see something like `const thing = require "thing"` around the internet, I think that's the older syntax from before the above syntax was decided upon.
```

Comment:
Require syntax is called [CommonJS](https://www.wikiwand.com/en/CommonJS), which was a server-side module definition pre-dating Node.js and made popular by Node. Eventually, around the time of ES6 a standard Javascript module format using the import/export statements was accepted. The reason for deviating from CommonJS (besides NIH syndrome) was that ES6 module format allows for better static guarantees (e.g. `import { find } from 'lodash'` can ensure that it doesn't need to include all of lodash in a bundle—which is where [rollup](http://rollupjs.org/) comes in for instance.

Also small note, but `require` is a plain ol' function so you can't omit the parens in vanilla JS so `const thing = require("thing")`